### PR TITLE
Implement heartbeat mechanism with JSON RPC protocol for WS  Agent

### PR DESCRIPTION
**Pull request by RomanNikitenko**
### What does this PR do?
- Implement heartbeat mechanism with JSON RPC protocol for WS  Agent on server side
- Tracks web socket connection to control ws agent state on client side. 
- Notifies client side by [WsAgentServerStoppedEvent](https://github.com/eclipse/che/blob/6e9825c699d34eef6b8bf36fc0194ff871cbee84/ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/workspace/event/WsAgentServerStoppedEvent.java#L22) when a connection is lost. 
- Prompts to restart a workspace when connection is lost, but workspace is running.

### What issues does this PR fix or reference?
#5935

Signed-off-by: Roman Nikitenko <rnikiten@redhat.com>

### CODE REVIEW

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+ * Tracks web socket connection to control ws agent state. Notifies client side by {@link
+ * WsAgentServerStoppedEvent} when the connection is lost. Prompts to restart a workspace when
+ * connection is lost, but workspace is running.
+ *
+ * @author Roman Nikitenko
+ */
+@Singleton
+public class WsAgentWebSocketConnectionTracker {
+  private static final int HEART_BEAT_PERIOD =
+      WS_AGENT_TRACK_CONNECTION_PERIOD + WS_AGENT_TRACK_CONNECTION_PERIOD / 2;
```
Comment (azatsarynnyy) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java]: WS_AGENT_TRACK_CONNECTION_PERIOD * 1.5

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  private void restartTimer() {
+    heartBeatTimer.cancel();
+    heartBeatTimer.schedule(HEART_BEAT_PERIOD);
+  }
+
+  private void onWsAgentConnectionIsLost() {
+    eventBus.fireEvent(new WsAgentServerStoppedEvent(devMachineName));
+
+    Workspace workspace = appContext.getWorkspace();
+    if (workspace == null || workspace.getStatus() != WorkspaceStatus.RUNNING) {
```
Comment (azatsarynnyy) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java]: IDE initialization has been improved a lot in Che 6.
Current workspace info set in `AppContext` during IDE initialization. And if IDE can't reach a workspace on loading - IDE initialization fails.
So, in Che 6, checking `workspace == null` is redundant since `appContext.getWorkspace()` can't return `null`.
If it isn't so, please ping me.

File: wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/Constants.java [hunk 3]
```diff
@@ -68,6 +68,13 @@
   public static final String SERVER_EXEC_AGENT_WEBSOCKET_REFERENCE = "exec-agent/ws";
 
   public static final String WS_AGENT_PORT = "4401/tcp";
+  public static final String WS_AGENT_TRACK_CONNECTION_SUBSCRIBE =
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/Constants.java]: This class is not related to wsagent. Consider moving these constants elsewhere.

**Review by dkulieshov — APPROVED** [review]

File: wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/WsAgentWebSocketConnectionKeeper.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * @author Roman Nikitenko
+ */
+@Singleton
+public class WsAgentWebSocketConnectionKeeper {
+  private final Set<String> endpointIds = newConcurrentHashSet();
+  private RequestTransmitter transmitter;
+  private RequestHandlerConfigurator configurator;
+
+  @Inject
+  public WsAgentWebSocketConnectionKeeper(
```
Comment (garagatyi) [inline wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/WsAgentWebSocketConnectionKeeper.java:54]: Can you elaborate on what will happen if the browser will be closed forcefully? Will it be unsubscribed ever? With the set grow? How would `sendRequest` method work if several entries from endpointIds are browsers that are gone?

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ * Tracks web socket connection to control ws agent state. Notifies client side by {@link
+ * WsAgentServerStoppedEvent} when the connection is lost. Prompts to restart a workspace when
+ * connection is lost, but workspace is running.
+ *
+ * @author Roman Nikitenko
+ */
+@Singleton
+public class WsAgentWebSocketConnectionTracker {
```
Comment (garagatyi) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java:50]: Consider adding tests

**Review by garagatyi — CHANGES_REQUESTED** [review]

File: wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/Constants.java [hunk 6]
```diff
@@ -68,6 +68,13 @@
...
+      "ws-agent/track/connection/subscribe";
+  public static final String WS_AGENT_TRACK_CONNECTION_UNSUBSCRIBE =
+      "ws-agent/track/connection/unsubscribe";
+  public static final String WS_AGENT_TRACK_CONNECTION_HEARTBEAT =
+      "ws-agent/track/connection/heartbeat";
+  public static final int WS_AGENT_TRACK_CONNECTION_PERIOD = 5_000; // milliseconds
```
Comment (azatsarynnyy) [inline wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/Constants.java]: Maybe `WS_AGENT_TRACK_CONNECTION_PERIOD_MILLIS` in order to make constant's name clearer and avoid redundant comment?

File: wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/WsAgentWebSocketConnectionKeeper.java (hunk 4 above)
Comment (dkulieshov) [reply wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/WsAgentWebSocketConnectionKeeper.java:54]: This is a common issue, should be fixed by this https://github.com/eclipse/che/issues/6177.

**Comment by RomanNikitenko** [comment]
I've tested this implementation and detected a bug.
Server side can be not responsive about 25 sec under certain conditions, for example when we have a mounted big project .
So I have to reconsider this approach to avoid some issues.

**Comment by dkulieshov** [comment]
@RomanNikitenko This happens probably due to usage of a common thread pool for all methods annotated with `@ScheduleRate`. Possible solution could be to use a separate single thread pool executor or something similar.

**Comment by RomanNikitenko** [comment]
@dkuleshov thank you very much! I'll try!

**Comment by RomanNikitenko** [comment]
I've updated the PR according to [comment](https://github.com/eclipse/che/pull/6969#issuecomment-341471305) and I need to test this implementation

File: wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/WsAgentWebSocketConnectionKeeper.java (hunk 4 above)
Comment (RomanNikitenko) [reply wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/WsAgentWebSocketConnectionKeeper.java:54]: I've added cleaning of unresponsive IDs

File: wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/Constants.java (hunk 3 above)
Comment (RomanNikitenko) [reply wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/Constants.java]: Constants are moved in shared module to have ability use them on client side and on server side

File: wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/Constants.java (hunk 6 above)
Comment (RomanNikitenko) [reply wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/Constants.java]: thank you, I've renamed constants and hope constants' names more clearer now

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java (hunk 5 above)
Comment (RomanNikitenko) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java:50]: I need to test current implementation and going to add tests

**Review by azatsarynnyy — APPROVED** [review]

**Review by garagatyi — APPROVED** [review]

**Comment by RomanNikitenko** [comment]
ci-build

**Comment by benoitf** [comment]
what's the status of this PR ?

**Comment by RomanNikitenko** [comment]
@benoitf I found server side can be not responsive about 25 sec under certain conditions, for example when we have a mounted big project. 
It does depend on implementation of heartbeat mechanism, but we have to investigate it and find the cause of the problem before merge.

**Comment by skabashnyuk** [comment]
Please change target branch of this pr to master branch.
